### PR TITLE
Transform service test suite

### PR DESCRIPTION
**Pull request by chamikaramj**
Adds a test suite for the Docker Composed based Transform Service.

Passing run: https://ci-beam.apache.org/job/beam_PostCommit_TransformService_Direct_PR/44/

Currently only includes Python BigTableIO integration tests, but we can add more tests to the suite in the future.

This addresses https://github.com/apache/beam/issues/27941.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by chamikaramj** [comment]
(not ready for review yet)

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by Abacn** [comment]
waiting on author

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

**Comment by chamikaramj** [comment]
R: @ahmedabu98

File: .test-infra/jenkins/job_PostCommit_TransformService_Direct.groovy [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      // Gradle goals for this job.
+      steps {
+        shell("echo \"*** RUN TRANSFORM SERVICE DIRECT USING PYTHON TRANSFORM_SERVICE_PYTHON_VERSION ***\"")
+        gradle {
+          rootBuildScriptDir(commonJobProperties.checkoutDir)
+          tasks(':sdks:python:test-suites:direct:xlang:transformServicePythonUsingJava')
+          commonJobProperties.setGradleSwitches(delegate)
+          switches '-PcompileAndRunTestsWithJava11'
+          switches "-Pjava11Home=${commonJobProperties.JAVA_11_HOME}"
+        }
```
Comment (ahmedabu98) [inline .test-infra/jenkins/job_PostCommit_TransformService_Direct.groovy:47]: FYI tests can run faster if we use Python wheels 
@Abacn helped me with that [here](https://github.com/apache/beam/pull/26491#issuecomment-1532321548)

File: sdks/python/test-suites/direct/xlang/build.gradle [hunk 2]
```diff
@@ -67,3 +69,20 @@ createCrossLanguageValidatesRunnerTask(
     "--endpoint localhost:${jobPort}",
   ],
 )
+
+createTransformServiceTask(
```
Comment (ahmedabu98) [inline sdks/python/test-suites/direct/xlang/build.gradle:73]: Can we add a comment here that these tests are run with Py38?

File: sdks/python/apache_beam/io/gcp/bigtableio_it_test.py [hunk 3]
```diff
@@ -50,6 +50,7 @@
 
 
 @pytest.mark.uses_gcp_java_expansion_service
+@pytest.mark.uses_transform_service
```
Comment (ahmedabu98) [inline sdks/python/apache_beam/io/gcp/bigtableio_it_test.py:53]: I may be missing something but how does this test suite make sure to build the necessary jars before running? For example in this test we need `:sdks:java:io:google-cloud-platform:expansion-service`

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run Seed Job

**Comment by chamikaramj** [comment]
Run TransformService_Direct PostCommit

File: sdks/python/test-suites/direct/xlang/build.gradle (hunk 2 above)
Comment (chamikaramj) [reply sdks/python/test-suites/direct/xlang/build.gradle:73]: Actually, these test can run with all Python versions. Updated.

(JavaUsingPython will be restricted to Python 3.8 I'll mention that later when we add such tests).

**Review by chamikaramj — COMMENTED** [review]
Thanks. PTAL.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 4]
```diff
@@ -2731,6 +2755,103 @@ class BeamModulePlugin implements Plugin<Project> {
...
+      }
+
+      def setupTask = project.tasks.register(config.name+"Setup", Exec) {
+        dependsOn ':sdks:java:container:'+javaContainerSuffix+':docker'
+        dependsOn ':sdks:python:container:py'+pythonContainerSuffix+':docker'
+        dependsOn ':sdks:java:transform-service:controller-container:docker'
+        dependsOn ':sdks:python:expansion-service-container:docker'
+        dependsOn ':sdks:java:expansion-service:container:docker'
+        dependsOn ":sdks:python:installGcpTest"
+        dependsOn project.project(':sdks:java:transform-service:launcher').shadowJar.getPath()
```
Comment (ahmedabu98) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:2807]: I think we need to also depend on task `initializeForDataflowJob` to set up wheels when we eventually add dataflow tests: https://github.com/apache/beam/blob/369265db4cd2aa294e6c4c4a4cfc1c8a9e4614c3/sdks/python/test-suites/dataflow/common.gradle#L41

```
if (usesDataflowRunner) {
  dependsOn ":sdks:python:test-suites:dataflow:py${project.ext.pythonVersion.replace('.', '')}:initializeForDataflowJob"
}
```

**Review by ahmedabu98 — COMMENTED** [review]
Just one comment. After that, this LGTM!

Most recent test pass (before spotless fix commit): https://ci-beam.apache.org/job/beam_PostCommit_TransformService_Direct_PR/44/

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 4 above)
Comment (chamikaramj) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:2807]: Done.

**Comment by chamikaramj** [comment]
Thanks!

**Comment by damccorm** [comment]
https://ci-beam.apache.org/job/beam_PostCommit_TransformService_Direct/ is also repeatedly failing

**Comment by chamikaramj** [comment]
https://github.com/apache/beam/pull/28039 should fix the "[PostCommit_TransformService_Direct/](https://ci-beam.apache.org/job/beam_PostCommit_TransformService_Direct/) failure

**Pull request merged**